### PR TITLE
[VAULT-3157] Move `mergeStates` utils from Agent to api module

### PR DESCRIPTION
**Pull request by vinay-gopalan**
This PR moves the utils `mergeStates`, `compareStates` from `api_proxy.go` in Agent to `client.go` in the API module. In addition, `ParseRequiredStates` from the `hashicorp/vault/vault` package is also moved to `client.go`. The `ParseRequiredStates` and `mergeStates` methods are also made public.

These methods need to be accessed by the Terraform Vault Provider in order to provide consistent X-Vault-Index header support.

### CODE REVIEW

File: api/client.go
Comment (tomhjp) [inline api/client.go:1157]:
```suggestion
func CompareReplicationStates(s1, s2 string) (int, error) {
```

States seems quite generic to me, and for a public method in our `api` package, won't give developers much intuition for what these methods are to do with unless they're already very familiar with Vault's internals. Is it worth making the method names a little more specific to help? Would love a second opinion on the suggestion though.

Comment (jasonodonnell) [reply api/client.go:1157]: I agree with Tom, and if you accept this please don't forget to change the documentation comment above it.

File: api/client.go
Comment (jasonodonnell) [inline api/client.go:1168]:
```suggestion
		return 0, fmt.Errorf("can't compare replication states with different ClusterIDs")
```

**Review by jasonodonnell — APPROVED** [review]
LGTM!

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: Thanks for the suggestion! Updated the method name

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: Updated, thanks!

File: api/client.go
Comment (tomhjp) [inline api/client.go:1183]:
```suggestion
func MergeReplicationStates(old []string, new string) []string {
```

Please could you add a function documentation comment as well, similar to CompareReplicationStates?

File: api/client.go
Comment (tomhjp) [inline api/client.go:1206]:
```suggestion
func ParseReplicationState(raw string, hmacKey []byte) (*logical.WALState, error) {
```

Also, does this one need to be public? I don't see it or `MergeStates` used in your TF provider PR.

Comment (vinay-gopalan) [reply api/client.go:1206]: Ah this was public when it was in Core and I didn't give a second deeper thought. We don't need `ParseReplicationState` to be public, but `MergeStates` does need to be since it is being used here https://github.com/hashicorp/terraform-provider-vault/pull/1170/files#diff-7e69de666c9006fa0a0fb8fe4e4f8c418acce7c555f69f8087d88d16b434f2c1R714

Comment (tomhjp) [reply api/client.go:1206]: Oh good point, I should have said `CompareReplicationStates` instead of `MergeStates`

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: Updated

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: Updated and made necessary methods private

File: api/client.go
Comment (calvn) [inline api/client.go:1240]:
```suggestion
		return nil, fmt.Errorf("invalid local index in state header: %w", err)
```

File: api/client.go
Comment (calvn) [inline api/client.go:1244]:
```suggestion
		return nil, fmt.Errorf("invalid replicated index in state header: %w", err)
```

File: api/client.go
Comment (calvn) [inline api/client.go:1218]: Not sure what would be the proper message here, but it'd be good to have it be different than the message on L1236

Maybe?
```suggestion
		return nil, fmt.Errorf("invalid full state header format")
```

**Review by calvn — APPROVED** [review]
Few minor nits, but otherwise looks good!

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: updated

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: updated

File: api/client.go
Comment (vinay-gopalan) [reply api/client.go]: updated

**Comment by ncabatoff** [comment]
Better make sure this works on Enterprise too.  ParseRequiredState is used in some ent-specific files.

**Comment by vinay-gopalan** [comment]
> Better make sure this works on Enterprise too. ParseRequiredState is used in some ent-specific files.

@ncabatoff the last spot that needs to be updated in Vault Enterprise is here in `core_util_ent.go`: https://github.com/hashicorp/vault-enterprise/blob/main/vault/core_util_ent.go#L1271

I'll open a PR on the `vault-enterprise` side updating that line once this work is merged in and available in the enterprise repo. Does that sound reasonable?

**Comment by ncabatoff** [comment]
> > Better make sure this works on Enterprise too. ParseRequiredState is used in some ent-specific files.
> 
> @ncabatoff the last spot that needs to be updated in Vault Enterprise is here in `core_util_ent.go`: https://github.com/hashicorp/vault-enterprise/blob/main/vault/core_util_ent.go#L1271
> 
> I'll open a PR on the `vault-enterprise` side by importing the `api` package in that file and updating that line once this PR is merged in and the changes available in the enterprise repo. Does that sound reasonable?

Yup, that's the usual practice.  Just don't want you to be caught unawares and scrambling to fix the build.

File: api/client.go
Comment (ncabatoff) [inline api/client.go:1209]: This needs to be re-exported so it's available to vault ent.

**Pull request merged**
